client/packages/programs/src/JsonForms/common/hooks/useAdditionalErrors.ts

```diff
@@ -0,0 +1,81 @@
+/**
+ * This hook provides methods for adding, updating or removing error objects
+ * from JSONForms' "additionalErrors" array.
+ *
+ * See https://jsonforms.io/docs/validation#external-validation-errors
+ *
+ * The update methods are added to the core "config" object to make them
+ * available to all renderer components, and the "additionalErrors" array is
+ * passed as a property on the main JsonForms component. Internally it is merged
+ * with its own validation errors, so can be accessed on the `errors` property
+ * of any control.
+ */
+
+import { useCallback, useState } from 'react';
+import { ErrorObject } from 'ajv';
+
+export const useAdditionalErrors = (
+  setError: ((error: string | false) => void) | undefined
+) => {
+  const [additionalErrors, setAdditionalErrors] = useState<ErrorObject[]>([]);
+
+  const addAdditionalError = useCallback(
+    (path: string, message: string) => {
+      const existing = additionalErrors.find(
+        error => error.instancePath === path
+      );
+      if (existing) {
+        if (existing?.message === message) return;
+        // Update existing error if message has changed
+        setAdditionalErrors(prevErrors => {
+          const newErrors = prevErrors.filter(
+            error => error.instancePath !== path
+          );
+          return [...newErrors, { ...existing, message }];
+        });
+        return;
+      }
+
+      // Add if new error
+      setAdditionalErrors(prevErrors => [
+        ...prevErrors,
+        {
+          instancePath: path,
+          message: message,
+          schemaPath: '',
+          keyword: 'custom',
+          params: {},
+        },
+      ]);
+
+      // This updates the form's overall "error" state, if it's defined
+      if (setError) setError(message);
```

Comment:
This is what allows the parent component to keep the "Save" button disabled even for the custom/additional errors. Previously, this didn't do this, as it wouldn't actually save the invalid values, but I think it's better still to not even allow the "Save" button to be clicked while there are any type of errors present.